wallet-cli/src/command/pretty.rs

```diff
@@ -3,32 +3,45 @@ use std::fmt;
 use std::string::String;
 
 use blockchain::{genesis, normal, types, Block, SscProof};
-use wallet_crypto::{address, hash, tx, util::hex};
+use wallet_crypto::{address, cbor, config, hash, hdwallet, redeem, tx, util::hex};
 
 use ansi_term::Colour;
 
 // Constants for the fmt::Display instance
 static DISPLAY_INDENT_SIZE: usize = 4; // spaces
 static DISPLAY_INDENT_LEVEL: usize = 0; // beginning starts at zero
 
-type AST = Vec<(Key, Val)>;
+type AST<'a> = Vec<(Key<'a>, Val<'a>)>;
 
-type Key = String;
+type Key<'a> = &'a str;
 
 // XXX: consider splitting into two mutually-recursive types (one with only terminals, one with only nonterminals)
 // TODO: extend with blockchain-specific constructors with color
-pub enum Val {
+pub enum Val<'a> {
     // terminals
     Raw(String),
-    Hash(types::HeaderHash),
+    Hash(types::HeaderHash), // XXX: consider naming this with a more specific meaning, as we'll probably have other hashes?
+    //// numbers
     Epoch(u32),
     SlotId(u32),
+    //// signatures
     BlockSig(normal::BlockSignature),
+    Signature(redeem::Signature),
+    //// actor ids
+    XPub(hdwallet::XPub),
     Stakeholder(address::StakeholderId),
 
     // recursive
-    List(Vec<Val>),
-    Tree(AST),
+    List(Vec<Val<'a>>),
+    Tree(AST<'a>),
+}
```

Comment:
The AST does not need to be **that** specific : one type for Hash (what ever algorithm) and one type for Signature is just as good.

```rust
pub enum Ast {
    // ...
    Hash(Vec<u8>),
    Signature(Vec<u8>),
    // ...
}
```

Reply:
the struct `normal::BlockSignature::{*}` includes two constructors which contain raw cbor and therefore aren't easily converted to `Signature(Vec<u8>)`.. I'd rather not convert them back to raw cbor bytes for pretty printing, so I'll use a dummy implementation which converts them to ascii bytes for now